microovn/ovn/commands.go

```diff
@@ -0,0 +1,150 @@
+package ovn
+
+import (
+	"errors"
+	"fmt"
+	"github.com/canonical/microcluster/state"
+	"github.com/lxc/lxd/shared"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+const defaultDBConnectWait = 30 //Default time to wait for connection to ovsdb
+
+// ovsdbSpec is a helper structure for precise identification of ovsdb databases. A lot of
+// ovn/ovs commands take path to either database file, database socket or process control socket
+// along with the database name. This structure can be used for such cases.
+type ovsdbSpec struct {
+	Path string
+	Name string
+}
+
+// OvsdbType is an enumeration of valid types of ovsdb databases which this package recognizes
+// and can execute commands against.
+type OvsdbType int
+
+const (
+	OvsdbTypeNB OvsdbType = iota
+	OvsdbTypeSB
+	OvsdbTypeSwitch
+)
+
+// newOvsdbSpec is a helper function that takes OvsdbType as an argument and generates
+// proper ovsdbSpec for given type.
+func newOvsdbSpec(dbType OvsdbType) (*ovsdbSpec, error) {
+	var dbSpec *ovsdbSpec
+	var err error
+
+	if dbType == OvsdbTypeNB {
+		dbSpec = &ovsdbSpec{
+			Path: filepath.Join(os.Getenv("SNAP_COMMON"), "run", "central", "ovnnb_db.sock"),
```

Comment:
The snap controls where OVS/OVN payload puts its files and sockets through environment variables, for example:
https://github.com/canonical/microovn/blob/a25b61d001db11d36800cbf1d39a4c4c7d0e493f/snapcraft/commands/central.start#L8-L12

We can't use that environment variable in the Go code directly, as the Go code manages multiple executables, each with their own environment. At the same time I don't feel comfortable also hard coding the same path in the Go code individually, would there be a way to provide these paths from a single source of truth?

At least we should start with a constant for the strings being repeated in the code?

Reply:
Agreed. They're are already sort of defined in here
https://github.com/canonical/microovn/blob/a25b61d001db11d36800cbf1d39a4c4c7d0e493f/microovn/ovn/paths.go#L9-L23
they are just not exporter.